tests/integration/test_steelthread.py

```diff
@@ -0,0 +1,196 @@
+import unittest
+import json
+import requests
+import pytest
+
+from nacl.signing import SigningKey
+import dwn_util
+import ssi_service_util
+import util
+
+# globals
+dwn_private_key = SigningKey.generate()
+dwn_did_public_key = dwn_util.get_did_key_from_bytes(
+    dwn_private_key.verify_key.encode()
+)
+create_did_response = None
+create_schema_response = None
+create_cred_manifest_response = None
```

Comment:
I think Gabe brings up a good point. In fact, fixtures can actually depend on other fixtures, and can be cached as well!

For example, see below a very broad approach. Notice that `(autouse=True)` is set for `cred_manifest`, so that all tests in scope use that fixture. I think even more things might be fixtures, like the object that's being written to the dwn collection.

Thoughts? 

```python
@pytest.fixture
def did():
  print("\nCreate a did for the issuer: ")

  with open(
      os.path.join(dirname, "fixtures/ssi-service/did-input.json"), "r"
  ) as file:
      jsonData = json.load(file)

  resp = http_request("PUT", "http://localhost:8080/v1/dids/key", jsonData)

  did_json = resp.json()

  assert resp.status_code == 201
  assert "did:" in did_json["did"]["id"]
  return did_json

@pytest.fixture
def schema(did):
    # Create a kyc schema for the manifest
    print("\nCreate a kyc schema for the manifest: ")

    issuer_did = did["did"]["id"]
    assert "did:" in issuer_did

    with open(
        os.path.join(dirname, "fixtures/ssi-service/kyc-schema-input.json"), "r"
    ) as file:
        jsonData = json.load(file)

    resp = http_request("PUT", "http://localhost:8080/v1/schemas", jsonData)

    create_schema_response = resp.json()

    assert resp.status_code == 201
    assert util.is_valid_uuid(create_schema_response["id"]) == True
    return create_schema_response

@pytest.fixture
def dwn_private_key():
    return SigningKey.generate()

@pytest.fixture
def dwn_public_key(dwn_private_key):
    return dwn_util.get_did_key_from_bytes(
        dwn_private_key.verify_key.encode()
    )

@pytest.fixture(autouse=True)
def cred_manifest(did, schema):
    # Create a credential manifest
    print("\nCreate a credential manifest: ")

    with open(
        os.path.join(dirname, "fixtures/ssi-service/manifest-input.json"), "r"
    ) as file:
        jsonData = json.load(file)

    issuer_did = did["did"]["id"]
    assert "did:" in issuer_did

    schema_id = schema["id"]
    assert schema_id is not None

    jsonData["issuerDID"] = issuer_did
    jsonData["outputDescriptors"][0]["schema"] = schema_id

    resp = http_request("PUT", "http://localhost:8080/v1/manifests", jsonData)

    create_cred_manifest_response = resp.json()

    assert resp.status_code == 201
    assert (
        util.is_valid_uuid(
            create_cred_manifest_response["credential_manifest"]["id"]
        )
        == True
    )
    return create_cred_manifest_response

class TestDWNRelayInstallProtocols:
    def test_install_protocols(self):
        # Configure Protocols for DWN-Relay
        print("\nConfigure Protocols for DWN-Relay: ")

        with open(
            os.path.join(dirname, "fixtures/dwn-relay/protocols-configure.json"), "r"
        ) as file:
            jsonData = json.load(file)

        resp = http_request("POST", "http://localhost:9000/", jsonData)

        assert resp.status_code == 200


class TestDWNRelayQueryManifestsDynamic:
    def test_query_manifests(self, dwn_private_key, dwn_did_public_key):
        # Query for the manifest from DWN-Relay
        message = dwn_util.create_collection_query_message(
            dwn_private_key, dwn_did_public_key
        )
        messages = {"messages": [message]}

        resp = http_request("POST", "http://localhost:9000/", messages)

        assert resp.status_code == 200
        assert (
            resp.json()["replies"][0]["entries"][0]["descriptor"]["schema"]
            == "https://identity.foundation/credential-manifest/schemas/credential-manifest"
        )


class TestDWNRelaySubmitCredApplicationDynamic:
    def test_submit_cred_app(self, cred_manifest):

        cred_manifest_id = cred_manifest["credential_manifest"]["id"]
        presentation_definition_id = cred_manifest[
            "credential_manifest"
        ]["presentation_definition"]["id"]

        with open(
            os.path.join(
                dirname, "fixtures/dwn-relay/collections-write-cred-app-input.json"
            ),
            "r",
        ) as file:
            cred_app_template = json.load(file)

        cred_app_template["credential_application"]["manifest_id"] = cred_manifest_id
        cred_app_template["credential_application"]["presentation_submission"][
            "definition_id"
        ] = presentation_definition_id

        vc = ssi_service_util.create_verifiable_credential()
        cred_app_template["verifiableCredentials"] = [vc["credentialJwt"]]

        cred_app_jwt = util.create_jwt(cred_app_template, dwn_private_key.encode())

        cred_app_jwt_object = {"applicationJwt": cred_app_jwt}

        message = dwn_util.create_collection_write_message(
            dwn_private_key, dwn_did_public_key, cred_app_jwt_object
        )
        messages = {"messages": [message]}

        resp = http_request("POST", "http://localhost:9000/", messages)

        # TODO: Fix the dynamic collections write message to have the correct DAG CID
        assert resp.status_code == 200

Reply:
Awesome, this works. Thanks for explaining! I'll implement it like this

Reply:
Yup much nicer now with fixtures